lib/handlers/authenticate-handler.js

```diff
@@ -0,0 +1,255 @@
+        response.set('WWW-Authenticate', 'Bearer realm="Service"');
+      }
+
+      if (!(e instanceof OAuthError)) {
+        throw new ServerError(e);
+      }
+
+      throw e;
+    });
+};
+
+/**
+ * Get the token from the header or body, depending on the request.
+ *
+ * "Clients MUST NOT use more than one method to transmit the token in each request."
+ *
+ * @see https://tools.ietf.org/html/rfc6750#section-2
+ */
+
+AuthenticateHandler.prototype.getToken = function(request) {
+  var headerToken = request.get('Authorization');
+  var queryToken = request.query.access_token;
+  var bodyToken = request.body.access_token;
+
+  if (!!headerToken + !!queryToken + !!bodyToken > 1) {
+    throw new InvalidRequestError('Invalid request: only one authentication method is allowed');
+  }
+
+  if (headerToken) {
+    return this.getTokenFromRequestHeader(request);
+  }
+
+  if (queryToken) {
+    return this.getTokenFromRequestQuery(request);
+  }
+
+  if (bodyToken) {
+    return this.getTokenFromRequestBody(request);
+  }
+
+  throw new UnauthorizedRequestError('Unauthorized request: no authentication given');
+};
+
+/**
+ * Get the token from the request header.
+ *
+ * @see http://tools.ietf.org/html/rfc6750#section-2.1
+ */
+
+AuthenticateHandler.prototype.getTokenFromRequestHeader = function(request) {
+  var token = request.get('Authorization');
+  var matches = token.match(/Bearer\s(\S+)/);
+
+  if (!matches) {
+    throw new InvalidRequestError('Invalid request: malformed authorization header');
+  }
+
+  return matches[1];
+};
+
+/**
+ * Get the token from the request query.
+ *
+ * "Don't pass bearer tokens in page URLs:  Bearer tokens SHOULD NOT be passed in page
+ * URLs (for example, as query string parameters). Instead, bearer tokens SHOULD be
+ * passed in HTTP message headers or message bodies for which confidentiality measures
+ * are taken. Browsers, web servers, and other software may not adequately secure URLs
+ * in the browser history, web server logs, and other data structures. If bearer tokens
+ * are passed in page URLs, attackers might be able to steal them from the history data,
+ * logs, or other unsecured locations."
+ *
+ * @see http://tools.ietf.org/html/rfc6750#section-2.3
+ */
+
+AuthenticateHandler.prototype.getTokenFromRequestQuery = function() {
+  throw new InvalidRequestError('Invalid request: do not send bearer tokens in query URLs');
+};
+
+/**
+ * Get the token from the request body.
+ *
+ * "The HTTP request method is one for which the request-body has defined semantics.
+ * In particular, this means that the "GET" method MUST NOT be used."
+ *
+ * @see http://tools.ietf.org/html/rfc6750#section-2.2
+ */
+
+AuthenticateHandler.prototype.getTokenFromRequestBody = function(request) {
+  if (request.method === 'GET') {
+    throw new InvalidRequestError('Invalid request: token may not be passed in the body when using the GET verb');
+  }
+
+  if (!request.is('application/x-www-form-urlencoded')) {
```

Comment:
@mjsalinger according to [RFC-6750](http://tools.ietf.org/html/rfc6750#section-2.2), the `x-www-form-urlencoded` is needed.
> The client MUST NOT use this method unless
   *all* of the following conditions are met:
   -  The HTTP request entity-header includes the "Content-Type" header
      field set to "application/x-www-form-urlencoded".